### PR TITLE
fix(gemma4): run SDPA in fp32 to avoid #2208 NaN on Hopper

### LINKED ISSUE #2208: Gemma4 NaN with SDPA
Some of the gemma4 models with lbs>=2 and attn_impl: sdpa throw a Nan loss value after few steps of fine-tuning. This appears in case of gemma4_2b.yaml, gemma4_4b.yaml, gemma4_31b_tp4.yaml and gemma4_31b_tp4_pp2.yaml. However gemma4_31b.yaml with just fsdp + ac runs fine, but produces Nan with TP=4 and TP4,PP2 cases.

The issue is also reproduced with 4b using AC.

Creating this issue to debug this further. Currently as a WAR these configs use eager attn_impl which runs fine without any NaN value (PR #2199). Also validated through loss curves for convergence.

More details on reproduction to be added soon.

### PR DESCRIPTION
**Pull request by aminehd**
## Summary

Fixes NVIDIA-NeMo/Automodel#2208. With `attn_implementation=sdpa` and `local_batch_size >= 2`, Gemma4 training throws a NaN loss after \~20 steps on Hopper (H100). This runs SDPA with fp32 query/key/value so attention is numerically stable, while keeping the memory-efficient SDPA kernel path (unlike eager, which materializes the full attention matrix).

## Root cause

The fused bf16 SDPA kernels are gradient-unstable for Gemma4 on Hopper: `grad_norm` goes NaN at step 22 and the loss follows at step 23. Eager avoids it by upcasting the softmax to fp32. Running SDPA with fp32 inputs gives the same numerical stability without leaving the fused/memory-efficient path.

## Reproduction

Using the reproducer from NVIDIA-NeMo/Automodel#2208 (`nvcr.io/nvidia/nemo-automodel:26.06`, 2×H100):

```
torchrun --nproc-per-node=2 examples/vlm_finetune/finetune.py \
  -c examples/vlm_finetune/gemma4/gemma4_2b.yaml \
  --step_scheduler.local_batch_size 2 --model.attn_implementation sdpa
```

* **Before:** `grad_norm` NaN @ step 22, loss NaN @ step 23.
* **After:** 30 steps clean, loss descending, no NaN.

Note: the NaN only reproduced on H100/Hopper — the same container and command ran clean on A100.

## Changes

* `nemo_automodel/components/models/gemma4_moe/sdpa_fp32.py` — `sdpa_fp32_attention_forward` (SDPA with fp32 q/k/v) and `enable_gemma4_sdpa_fp32()`.
* `nemo_automodel/components/models/gemma4_moe/model.py` — install it from `from_pretrained` when `attn_implementation=sdpa`.

### CODE REVIEW

**Comment by athitten** [comment]
@aminehd thank you for the PR! Mind running gemma4 2b for say 100 steps similar to [https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/vlm_finetune/gemma4/gemma4_2b.yaml](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/vlm_finetune/gemma4/gemma4_2b.yaml) but with SDPA and share the loss curve ?

**Comment by aminehd** [comment]
Hi @athitten, thanks for the review. I ran gemma4 2b for 100 steps with SDPA and the fp32 fix ([this script](https://github.com/aminehd/automodel-repro/blob/main/examples/loss_curve_2606.py)), and here is the loss curve:

gemma4_2b SDPA loss curve — fp32 fix stable over 100 steps vs baseline NaN at step 22

**Comment by akoumpa** [comment]
Hi @aminehd  could you take a look at the linting error https://github.com/NVIDIA-NeMo/Automodel/actions/runs/29964982812/job/89074389585?pr=3141 and run the linter? once ready, please let me know and i can help rerun ci. thanks

**Comment by aminehd** [comment]
Thanks @akoumpa , sorry for missing the linting part. I ran `ruff format` to fix lint issues. afterward `ruff format --check` and `ruff check` both passed without changing files.
Updated the commit and pushed to branch in this cr

**Comment by athitten** [comment]
@aminehd thank you for loss curve! Would you mind adding some unit tests for this. CodeCov is likely going to complain since there are no tests for `nemo_automodel/components/models/gemma4_moe/sdpa_fp32.py` and its usage ?

**Comment by athitten** [comment]
> Hi @athitten, thanks for the review. I ran gemma4 2b for 100 steps with SDPA and the fp32 fix ([this script](https://github.com/aminehd/automodel-repro/blob/main/examples/loss_curve_2606.py)), and here is the loss curve:
> 
> gemma4_2b SDPA loss curve — fp32 fix stable over 100 steps vs baseline NaN at step 22

Thanks again, if you have throughput, memory numbers (for the first 20 steps) comparing the two feel free to add them too. It can be useful!

**Comment by aminehd** [comment]
Hi @athitten , 
Yeah I will add some tests for sdpa_fp32.py and its usage.
I'll run the experiments again and capture  throughput, memory numbers for the two as well

**Comment by aminehd** [comment]
Hi @athitten @akoumpa

Per-step throughput (tokens/sec) and peak memory (SDPA+fp32 vs eager, gemma4_2b, lbs=2, 2×H100, activation checkpointing on, first 20 steps):

per-step throughput (tok/s) and peak memory — fp32 fix vs eager

Memory is identical between the two at every step (the two lines overlap in the top panel); fp32 SDPA has higher throughput.

<details>
<summary>Per-step numbers</summary>

| step | fix mem (GiB) | fix tok/s | eager mem (GiB) | eager tok/s |
|---|---|---|---|---|
| 0 | 37.57 | 254 | 37.57 | 253 |
| 1 | 46.19 | 2791 | 46.19 | 1983 |
| 2 | 46.19 | 2842 | 46.19 | 2626 |
| 3 | 46.19 | 3062 | 46.19 | 3059 |
| 4 | 46.19 | 2797 | 46.19 | 2204 |
| 5 | 52.72 | 3323 | 52.73 | 3453 |
| 6 | 46.20 | 3219 | 46.20 | 2594 |
| 7 | 46.20 | 2862 | 46.20 | 2416 |
| 8 | 46.19 | 2394 | 46.19 | 2021 |
| 9 | 46.19 | 2541 | 46.19 | 2091 |
| 10 | 46.19 | 2786 | 46.19 | 2165 |
| 11 | 50.30 | 2288 | 50.30 | 2207 |
| 12 | 46.20 | 2986 | 46.20 | 2278 |
| 13 | 46.20 | 2989 | 46.20 | 2444 |
| 14 | 46.19 | 3413 | 46.19 | 2938 |
| 15 | 46.19 | 2952 | 46.19 | 2385 |
| 16 | 46.20 | 2925 | 46.20 | 2362 |
| 17 | 46.19 | 2568 | 46.19 | 2102 |
| 18 | 46.19 | 2469 | 46.19 | 2081 |
| 19 | 47.46 | 3429 | 47.46 | 3138 |

</details>

I added tests as well. Quick question on the test layout. I put the `sdpa_fp32.py` tests and the `from_pretrained` usage tests in one file. Happy to split the usage tests into `test_gemma4_model.py` if you'd rather.

**Review by athitten — APPROVED** [review]
Thanks again for the PR @aminehd  and for addressing code review comments!

**Comment by aminehd** [comment]
Thanks a lot @athitten for providing the reproduction steps and help with the reviews
